Studies/source/PuritySignalStudy.cxx

```diff
@@ -0,0 +1,306 @@
+                index_matched_jet.push_back(counter);
+            ++counter;
+        }
+
+        // find common elements between the indexes that refer to the two best indexes after ordering
+        // and the indexes that refer to the elements with match
+        auto common_indexes = intersection(index_matched_jet, best_two);
+        anaData.jet_matches(ToString(jet_ordering)).Fill(common_indexes.size());
+
+        return common_indexes.size();
+    }
+
+    size_t calculateMatchesTaus(const Event& event, TauIdDiscriminator tau_id_discriminator, Channel channel, bool is_descending_order = true)
+    {
+        std::vector<size_t> ordered_indexes = orderTaus(event, tau_id_discriminator, is_descending_order);
+        std::vector<size_t> best_two = {ordered_indexes.at(0)};
+
+        if(args.channel() == Channel::TauTau)
+            best_two.push_back(ordered_indexes.at(1));
+
+        // find indexes that refer to the elements with match
+        std::vector<size_t> index_matched_tau;
+        int counter = 0;
+        for(auto i : event.lep_genTauIndex){
+            if(i != 10) // 10 is the defult value of this branch on the ntuple
+                index_matched_tau.push_back(counter);
+            ++counter;
+        }
+
+        // find common elements between the indexes that refer to the two best indexes after ordering
+        // and the indexes that refer to the elements with match
+        auto common_indexes = intersection(index_matched_tau, best_two);
+
+        anaData.lep_matches(ToString(tau_id_discriminator)).Fill(common_indexes.size());
+
+        return common_indexes.size();
+    }
+
+    std::vector<size_t> getBestTwoJetsIndexes(std::vector<analysis::jet_ordering::JetInfo<LorentzVector>> input)
+    {
+        std::vector<size_t> output;
+        int ctr = 0;
+
+        for (auto input_i : input) {
+            output.push_back(input_i.index);
+            ++ctr;
+            if (ctr==2)
+                break;
+        }
+       return output;
+    }
+
+private:
+
+    std::vector<analysis::jet_ordering::JetInfo<LorentzVector>> orderJets(const Event& event, JetOrdering jet_ordering/*, Period period*/)
+    {
+        std::vector<analysis::jet_ordering::JetInfo<LorentzVector>> jet_info_vector;
+        BTagger bTagger(Period::Run2017,jet_ordering);
+
+        for(size_t jet_index = 0; jet_index < event.jets_p4.size(); ++jet_index)
+            jet_info_vector.emplace_back(static_cast<LorentzVectorE>(event.jets_p4.at(jet_index)), jet_index, bTagger.BTag(event,jet_index));
+
+
+        return jet_ordering::OrderJets(jet_info_vector, false, 20, 2.1);
+    }
+
+    std::vector<size_t> orderTaus(const Event& event, TauIdDiscriminator tau_id_discriminator, bool is_descending_order = true)
+    {
+       std::vector<float> raw_values;
+       for(size_t lep_index = 0; lep_index < event.lep_p4.size(); ++lep_index){
+           if (static_cast<LegType>(event.lep_type.at(lep_index)) == analysis::LegType::tau ) {
```

Comment:
```
ntuple::TupleLepton lepton(event, lep_index);
const float raw = lepton.type() == LegType::tau ? lepton.GetRawValue(tau_id_discriminator) : 0;
raw_values.push_back(raw);
```

Reply:
please, review the code: raw_values should have the same size as event.lep_p4, otherwise, indices will be wrong.